actions/workflows/create_vm_role.yaml

```
@@ -1,100 +1,127 @@
-        value: "{{ec2_instance_private_dns}}"
-        zone: "{{dns_zone}}"
-      publish:
-        cname: "{{hostname}}.{{dns_zone}}"
-      on-success: "get_distro"
-    -
-      name: "get_distro"
-      ref: "st2cd.get_distro"
-      params:
-        hosts: "{{ec2_instance_private_ip}}"
-      on-success: "set_hostname"
-    -
-      name: "set_hostname"
-      ref: "st2cd.set_hostname"
-      params:
-        hosts: "{{ec2_instance_private_ip}}"
-        hostname: "{{hostname}}.{{dns_zone}}"
-        distro: "{{get_distro[ec2_instance_private_ip].stdout}}"
-        timeout: 300
-      on-success: "reboot"
-    -
-      name: "reboot"
-      ref: "core.remote_sudo"
-      params:
-        hosts: "{{ec2_instance_private_ip}}"
-        cmd: "shutdown -r +1 &"
-      on-success: "wait_for_ssh_post_reboot"
-    -
-      name: "wait_for_ssh_post_reboot"
-      ref: "linux.wait_for_ssh"
-      params:
-        hostname: "{{ec2_instance_private_ip}}"
-        username: "stanley"
-        keyfile: "{{keyfile}}"
-        ssh_timeout: 30
-        retries: 10
-      on-success: "puppet_bootstrap"
-    -
-      name: "puppet_bootstrap"
-      ref: "st2cd.puppet_bootstrap"
-      params:
-        hosts: "{{ec2_instance_private_ip}}"
-        role: "{{role}}"
-  default: "get_subnet_id"
+version: '2.0'
+
+st2cd.create_vm_role:
+    description: Create VM role on AWS
+    type: direct
+    input:
+        - instance_type
+        - environment
+        - key_name
+        - keyfile
+        - hostname
+        - dns_zone
+        - role
+        - creator
+        - purpose
+        - distro
+        - instance_id
+    tasks:
+        get_subnet_id:
+            action: st2.kv.get
+            input:
+                key: <% $.environment %>_subnet
+            on-success:
+                - get_ami
+
+        get_ami:
+            action: st2.kv.get
+            input:
+                key: ami_<% $.distro %>
+            on-success:
+                - run_instance
+
```

Checking if cname exists would be a really good first task.  At least then we fail fast.